### DISCUSSION TITLE
[Feature]: Kalender außerhalb von Lotta zur Verfügung stellen

### DISCUSSION THREAD

**Discussion by wie0815** [lotta-schule/web#231] [Ideen]
Das Ticket erfordert, dass zuerst #170 erledigt wird.

In Lotta erstellte Kalender sollten nicht nur innerhalb von Lotta in der Marginale zur Verfügung stehen, sondern auch außerhalb.

### Wofür?
Zu klären

### Möglichkeiten
- Download einer iCal Datei (nach [RFC5545](https://datatracker.ietf.org/doc/html/rfc5545) . Diese beinhaltet Termine zum Zeitpunkt des Downloads.
- Anbieten eines [CalDAV](https://en.wikipedia.org/wiki/CalDAV)-Endpunktes. Kalendereinträge würden damit automatisch in den Client heruntergeladen werden und auch in dem Client bearbeitet werden können

#### Probleme für Möglichkeit eines Downloads einer statischen iCal-Datei
Übersichtlicher Aufwand.
Man könnte den Download als Funktionserweiterung der Marginale anbieten. Dann wäre die Frage, wer die Datei herunterladen darf, geklärt: Wer Leserechte auf die Marginale hat, sieht dort den Link und bekommt über diesen als angemeldeter Nutzer auch die Datei. Problem wäre die Frage, ob eine statische Funktion wirklich sinnvoll wäre und welcher Nutzen damit abgedeckt werden soll

#### Probleme bei CalDAV
- Wir müssten zuerst einen WebDAV Server implementieren. Das alleine ist recht aufwendig. Eigenimplementierung oder Nutzen [vorhandener Server](https://devguide.calconnect.org/CalDAV/Server-Implementations/) wäre möglich
- Information, wer den Kalender lesen und wer ihn bearbeiten kann müssten am Kalender geklärt werden
- Authentifizierung müsste geklärt werden. Am einfachsten wäre sich über seine Lotta-Zugangsdaten anzumelden. Besser wäre das Anbieten von [OAuth-Provider](https://oauth.net/2/) oder einer Art "App-spezifische Passwörter" um zu vermeiden dass der Nutzer seine Zugangsdaten dubiosen Kalenderdiensten übergibt / wir uns später Features wie [multi-Factor-Auth](https://en.wikipedia.org/wiki/Multi-factor_authentication) oder [Passkeys](https://en.wikipedia.org/wiki/WebAuthn) offen halten
- CalDAV hätte dafür den Vorteil, dass wir dann schon eine Infrastruktur für [WebDAV](https://en.wikipedia.org/wiki/WebDAV) hätten die wir theoretisch auch für andere Features wie Einbindung von Lotta-Dateien ins Betriebssystem

**Comment by ptitmouton** [comment]
@wie0815 Ich hoffe ich habe deine Erwähnung des iCal-Formats richtig verstanden. Du möchtest den Kalender außerhalb von Lotta verfügbar machen.

Die Sache ist recht umfangreich und ich sehe in keinem Fall dass das alles in einem einzigen Rutsch erledigt wird.

Wir müssten erstmal klären was wir damit genau ermöglichen wollen.

**Comment by wie0815** [comment]
Ok. ... wir müssen darüber reden, denn ich kann Vor- und Nachteile bzw. Schwierigkeiten nicht nachvollziehen (#Fachsprache), d.h. ich kann Deiner Beschreibung nicht folgen @ptitmouton.

**Comment by ptitmouton** [comment]
Also nach persönlicher Diskussion haben wir uns ja auf folgendes geeinigt:

Kalender, die in Lotta erstellt wurden, sollen per öffentlich zugänglicher ics-Datei im iCal Format zur Verfügung gestellt werden. 

Die meisten Kalender-Programme unterstützen die Einbindung dann als ‚Kalenderabo‘ oder ‚externer Kalender‘ oder sowas in der Art. 

Hier würde sich quasi nur noch die Frage stellen, wie wir den Link Nutzern zur Verfügung stellen möchten @wie0815 @BillyTheKing . 

Folgende Optionen fallen mir ein:

*über die Marginale*
In der Marginale ist ein Icon zu sehen, dass ein Abonnement repräsentiert. Bei mehreren Kalendern innerhalb einer Marginale könnte es neben dem Kalendernamen stehen, bei einem einzigen Kalender müsste es anders positioniert werden.
Klickt man das Icon an, öffnet sich ein Dialog, der ein Textfeld mit dem Link beinhaltet. Daneben könnte ein Kopier-Button sein, der den Link bei Klick in die Zwischenablage kopiert. In einem kurzen Beschreibungstext darüber wird erklärt, dass jeder, der den angezeigten öffentlich zugänglichen Link in seiner Kalendersoftware abonniert, lesenden Zugriff auf die Termine hat.

Vorteil:
Für den Endnutzer einfach zugänglich und an der Stelle, an der er üblicherweise den Kalender konsumiert

Nachteil:
Der Kalender muss auch tatsächlich in einer Marginalen eingebunden werden um überhaupt entdeckt zu werden. Außerdem muss der Anwender wissen, dass hier eine Aktion möglich ist. Ist vermutlich nicht sehr selbsterklärend. 


*Der Link wird im Admin-Bereich an der Stelle angezeigt, an der auch der Kalender erstellt wird*
in der Kalenderansicht im Adminbereich - wie auch immer sie genau aussehen wird - könnte dort, wo man Kalender bzw. dessen Termine einstellt, ein Button „Kalender abonnieren“ verfügbar sein. Klickt man ihn an, öffnet sich der schon oben beschriebene Dialog. So hätte ein Administrator Zugriff zum Link und könnte ihn an berechtigte Personen weitergeben.

Vorteil:
Die Verteilung des Links würde aktiv durch einen Administrator geschehen, der so eine minimale Kontrolle über die Verbreitung des Links hätte, und auch aktiv auf Nutzer zugehen könnte und ihnen ihre Möglichkeit, den Kalender zu abonnieren, bei Übergabe des Links näher bringen könnte.

Nachteil:
Der Zugang für Berechtigte erfolgt nur indirekt, sie wären davon abhängig dass ein Admin ihnen den Link gibt. Eventuell würden viele potentielle Nutzer gar nicht von der Möglichkeit erfahren, Kalender abonnieren zu können.






Was meint ihr? Andere Ideen?
Ich persönlich würde momentan dazu tendieren, beide Möglichkeiten einzubauen. 
vielleicht übersehe ich aber auch eine einfachere und vor Allem zugänglichere Lösung?

**Comment by wie0815** [comment]
@ptitmouton: Ich bin sehr dafür, dass wir alle Entscheidungsgewalt bei den Administratoren der Lottas lassen. Grundsätzlich sollten die Admins den Kalender-Link einsehen, kopieren und weitergeben können. In dem Kontext sollte der Link im Admin-Bereich gut sichtbar positioniert werden.

Vielleicht können wir an der Stelle (im Admin-Bereich) ein Auswahlfeld setzen: "**Kalenderadresse in der Marginale anzeigen**". Somit hätten wir beide Möglichkeiten im Angebot.
